### PR TITLE
521 notice parcing failure

### PR DESCRIPTION
**Pull request by RollingHog**
#521 adding notice for failed parsing
TESTS NOT FIXED, WAITING FOR OVERALL APPROVAL/CHANGES

UI changes:
- if there are `conf/runners` .json files script-server cannot parse, red group called `` will appear in main scripts list

Backend changes:
- `list_configs()` now returns mixed `List[ Union[ShortConfig, GetShortConfigFailedError] ]` list which is filtered in `GetScripts()`
- `server.py` `GetScripts()` endpoint - now returns JSON `{'scripts': scripts, 'failed': failed}` where `'failed'` is "failed parsing" 

Frontend changes (I had no idea what I'm doing):
- created `ScriptFailedGroup.vue` closely resembling `ScriptListGroup.vue` to show list of non-interactive unparsed files

### CODE REVIEW

**Comment by RollingHog** [comment]
And I can't see exactly which error in pretty big Travis feed fails. Never used tests in Python, sry.

**Comment by bugy** [comment]
Hi @RollingHog, thanks for your changes. I would see it implemented a bit differently:
- Instead of returning completely different type from `list_configs`, I think we should just return existing short_config type with a new field "parsing_failed: true". Returning 2 completely different types from a method is more difficult to read and maintain
- if a script file contains text: "allows_users", then it shouldn't be listed at all (otherwise users might see scripts, which are not intended for them)
- In all other places in config_service, if we are loading extended model for a config, we should check this flag as well. I.e. if a user tries to load and display such a script, we would show proper error
- On UI, instead of using a separate group, I would suggest to show the script as it is, but: make the text red and add error icon to it
- Admin UI should be adjusted as well, i.e. the config should be listed, but shouldn't be clickable, and shown as read


Regarding the tests, you can run them using the follwing command from `src` folder:
`python3 -m unittest discover -s tests -p "*.py" -t .`

**Comment by RollingHog** [comment]
O-okay.

**Comment by RollingHog** [comment]
Will update this during work

> - Instead of returning completely different type from list_configs, I think we should just return existing short_config type with a new field "parsing_failed: true". 
> - if a script file contains text: "allows_users", then it shouldn't be listed at all (otherwise users might see scripts, which are not intended for them)
> - In all other places in config_service, if we are loading extended model for a config, we should check this flag as well. I.e. if a user tries to load and display such a script, we would show proper error
> - On UI, instead of using a separate group, I would suggest to show the script as it is, but: make the text red and add error icon to it
> - Admin UI should be adjusted as well, i.e. the config should be listed, but shouldn't be clickable, and shown as read

**Comment by RollingHog** [comment]
Question emerged:
* On `allowed_users`: should I regexp JSON server-side and return something like "possibly restricted script" instead of its real filename for non-admin users?

**Comment by bugy** [comment]
> On allowed_users: should I regexp JSON server-side and return something like "possibly restricted script" instead of its real filename for non-admin users?

Yes please, please check the file (I think regex is not needed, just check, that the file content contains "allowed_users" string)
However, do not show anything to non-admin users at all. The file might be supposed to be hidden from them

**Comment by RollingHog** [comment]
> For the stacktrace, I would suggest to log it. It will cause some noise, but it's easier to investigate having traceback

I meant only for "cannot parse" error of course. Still log?

And about files. I think a note "some file with restricted access is malformed" will be good, wont it? Or you have some reason to fully hide this info?

**Comment by bugy** [comment]
> I meant only for "cannot parse" error of course. Still log?

yes, it can still provide some hints on what's wrong with the file.

> I think a note "some file with restricted access is malformed" will be good, wont it? Or there is some reason to fully hide this info?

Not for normal users. Any restricted information should be hidden from users, to avoid any leaks. Even if it's only a file name

**Comment by RollingHog** [comment]
> Even if it's only a file name

No no no! What I planned to do was replacing filename with "file with possibly restricted access" server-side. So users see that something is wrong (and can eg notify admin) but don't know what exactly.

**Comment by bugy** [comment]
> What I planned to do was replacing filename with "file with possibly restricted access" server-side.

Ah, I see. I thought, that you wanted to display filename and show the error
May be in this case, it would be just enough to show how many files failed to load, e.g. in the script list UI it would look like this:
- Script 1
- Script 2
- Script 3
- (red) N files failed to load

**Comment by RollingHog** [comment]
Did all. Now trying to understand testing command output to fix the tests.

File: web-src/src/main-app/components/scripts/ScriptListItem.vue [hunk 1]
```diff
@@ -96,26 +105,45 @@ export default {
...
 
+/* state: finished */
+
 .scripts-list .collection-item .menu-item-state.finished > .check-icon {
   display: block;
 }
 
-.scripts-list .collection-item .menu-item-state.finished > .preloader-wrapper {
+.scripts-list .collection-item .menu-item-state.finished > .preloader-wrapper,
+.scripts-list .collection-item .menu-item-state.finished > .failed-icon {
```
Comment (bugy) [inline web-src/src/main-app/components/scripts/ScriptListItem.vue:131]: I think there is no need to duplicate this `display:none` multiple times. 
You can define it once, smth like:
```
.scripts-list .collection-item .menu-item-state > .failed-icon {
  display: none;
}
```

And it should be enough.
I do not know why I copied it multiple times for check icon and preload wrapper.

File: web-src/src/main-app/store/scriptConfig.js [hunk 2]
```diff
@@ -333,8 +334,14 @@ function reconnect(state, internalState, commit, dispatch, selectedScript) {
             logError(error);
 
             if (error instanceof SocketClosedError) {
+              
+                if (error.code === 422) {
+                  commit('SET_ERROR', `${CANNOT_PARSE_ERROR_PREFIX} "${selectedScript}"`);
+                  return;
+                }
```
Comment (bugy) [inline web-src/src/main-app/store/scriptConfig.js:341]: I would suggest to remove custom handling here. Script server can start sending 422 for other cases as well, so always showing "parsing error" is not good. We should always rely on `error.reason`, which is provided by the server

File: src/model/script_config.py [hunk 3]
```diff
@@ -27,7 +27,16 @@ def __init__(self):
...
         self.group = None
-
+        self.parsing_failed = False
+
+def create_failed_ShortConfig(path):
+    failed_short_config = ShortConfig()
+    failed_short_config.name = path
+    with open(path) as f:
+      if '"allowed_users"' in f.read():
+          failed_short_config.name = 'file with possibly restricted access'
```
Comment (bugy) [inline src/model/script_config.py:37]: I'm not sure that having 2 configs with the same name would work fine for script server.

File: web-src/src/admin/components/scripts-config/ScriptsList.vue [hunk 4]
```diff
@@ -37,7 +49,7 @@ export default {
   computed: {
     ...mapState('scripts', {
       scripts: state => {
-        return state.scripts ? state.scripts.map(s => ({name: s, path: encodeURIComponent(s)})) : []
+        return state.scripts ? state.scripts.map(s => ({name: s.name, path: encodeURIComponent(s.name), parsingFailed: s.parsingFailed})) : []
```
Comment (bugy) [inline web-src/src/admin/components/scripts-config/ScriptsList.vue:52]:
```suggestion
        return state.scripts ? state.scripts.map(s => ({
              name: s.name, 
              path: encodeURIComponent(s.name), 
              parsingFailed: s.parsingFailed
              })) 
              : []
```

**Review by bugy — CHANGES_REQUESTED** [review]

File: web-src/src/admin/components/scripts-config/ScriptsList.vue [hunk 5]
```diff
@@ -62,4 +74,13 @@ export default {
 .add-script-btn {
   margin-bottom: 1em;
 }
+
+.scripts-list .collection-item.parsing-failed {
+  color: var(--error-color);
+  /* I didn't find any good ways to disable a <router-link> */
+  /* pointer-events: none; */
```
[the remark this answers is not in the document]
Comment (RollingHog) [reply web-src/src/admin/components/scripts-config/ScriptsList.vue:81]: Yes, I tried, didn't work.

File: src/config/config_service.py [hunk 6]
```diff
@@ -364,3 +381,10 @@ def __init__(self, message):
 class InvalidAccessException(Exception):
     def __init__(self, message=None):
         super().__init__(message)
+
+
+class BadConfigFileException(Exception):
+    HTTP_CODE = 422
```
[the remark this answers is not in the document]
Comment (RollingHog) [reply src/config/config_service.py:388]: Added in proper usecases

File: web-src/src/main-app/components/scripts/ScriptListItem.vue [hunk 7]
```diff
@@ -96,26 +105,45 @@ export default {
...
+.scripts-list .collection-item .menu-item-state.executing > .check-icon,
+.scripts-list .collection-item .menu-item-state.executing > .failed-icon {
   display: none;
 }
 
 .scripts-list .collection-item .menu-item-state.executing > .preloader-wrapper {
   display: block;
 }
 
+/* state: finished */
```
[the remark this answers is not in the document]
Comment (RollingHog) [reply web-src/src/main-app/components/scripts/ScriptListItem.vue:124]: I think this is still useful due to rather strange CSS, but as you wish.

File: web-src/src/main-app/components/scripts/ScriptListItem.vue (hunk 1 above)
Comment (RollingHog) [reply web-src/src/main-app/components/scripts/ScriptListItem.vue:127]: Not quite sure what do you mean, sorry.

File: src/config/config_service.py [hunk 8]
```diff
@@ -191,7 +196,10 @@ def load_script(path, content):
                     return None
 
                 return short_config
-            except:
+            except json.decoder.JSONDecodeError:
+                failed_short_config = create_failed_ShortConfig(path)
```
[the remark this answers is not in the document]
Comment (RollingHog) [reply src/config/config_service.py]: Fixed.

File: src/model/script_config.py (hunk 3 above)
Comment (RollingHog) [reply src/model/script_config.py]: Checked, it really causes some problems. Added path hash.

File: src/config/config_service.py [hunk 9]
```diff
@@ -191,7 +196,10 @@ def load_script(path, content):
                     return None
 
                 return short_config
-            except:
+            except json.decoder.JSONDecodeError:
```
[the remark this answers is not in the document]
Comment (RollingHog) [reply src/config/config_service.py:199]: Isn't logging already present there?.. Line 200

File: src/config/config_service.py [hunk 10]
```diff
@@ -364,3 +381,10 @@ def __init__(self, message):
 class InvalidAccessException(Exception):
     def __init__(self, message=None):
         super().__init__(message)
+
+
+class BadConfigFileException(Exception):
+    HTTP_CODE = 422
+    VERBOSE_ERROR = 'Cannot parse script config file'
```
[the remark this answers is not in the document]
Comment (RollingHog) [reply src/config/config_service.py:389]: So you mean message=VERBOSE_ERROR?

**Review by bugy — APPROVED** [review]

**Comment by bugy** [comment]
@RollingHog I'm so sorry, I totally forgot about this PR. I thought that we finished it. 
I'm merging it to master now, thanks for your work!

**Comment by RollingHog** [comment]
Well, I hope it helped.

**Pull request merged**